spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java

```diff
@@ -253,46 +258,59 @@ public ArrayBasedMapData read(Decoder decoder, Object reuse) throws IOException
     }
   }
 
-  static class StructReader implements ValueReader<InternalRow> {
-    private final ValueReader<?>[] readers;
+  static class StructReader extends ValueReaders.StructReader<InternalRow> {
+    private final int numFields;
 
-    private StructReader(List<ValueReader<?>> readers) {
-      this.readers = new ValueReader[readers.size()];
-      for (int i = 0; i < this.readers.length; i += 1) {
-        this.readers[i] = readers.get(i);
-      }
+    protected StructReader(List<ValueReader<?>> readers, Types.StructType struct, Map<Integer, ?> idToConstant) {
+      super(readers, struct, idToConstant);
+      this.numFields = readers.size();
     }
 
-    ValueReader<?>[] readers() {
-      return readers;
+    @Override
+    protected InternalRow reuseOrCreate(Object reuse) {
+      if (reuse instanceof GenericInternalRow && ((GenericInternalRow) reuse).numFields() == numFields) {
+        return (InternalRow) reuse;
+      }
+      return new GenericInternalRow(numFields);
     }
 
     @Override
-    public InternalRow read(Decoder decoder, Object reuse) throws IOException {
-      GenericInternalRow row = new GenericInternalRow(readers.length);
-      if (decoder instanceof ResolvingDecoder) {
-        // this may not set all of the fields. nulls are set by default.
-        for (Schema.Field field : ((ResolvingDecoder) decoder).readFieldOrder()) {
-          Object value = readers[field.pos()].read(decoder, null);
-          if (value != null) {
-            row.update(field.pos(), value);
-          } else {
-            row.setNullAt(field.pos());
-          }
-        }
+    protected Object get(InternalRow struct, int pos) {
+      return null;
```

Comment:
Should we return the obj at pos here?

Reply:
The previous version of this code didn't implement reuse, probably because it needs the Spark type to correctly fetch the value. Rather than fix that in this PR, we can implement reuse in another one.

Reply:
Sounds good!